Allow storage() to work on python tensor subclasses, but error on future data accesses

bdhirsh opened the pull request:
This was discussed in feedback from the original version of my "reorder proxy/fake" PR. This PR allows calls to `tensor.untyped_storage()` to **always** return a python storage object to the user. Previously, we would error loudly if we detected that the storage had a null dataptr.

Instead, I updated the python bindings for the python storage methods that I saw involve data access, to throw an error later, only if you try to access those methods (e.g. `storage.data_ptr()` will now raise an error if the data ptr is null).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105236
* #104974
* #104483
* #106406
* #106405
* #106404
* #104482
* #107415
* #107416
* __->__ #107417

bdhirsh committed:
Allow storage() to work on python tensor subclasses, but error on future data accesses

[ghstack-poisoned]

bdhirsh committed:
Update on "Allow storage() to work on python tensor subclasses, but error on future data accesses"

This was discussed in feedback from the original version of my "reorder proxy/fake" PR. This PR allows calls to `tensor.untyped_storage()` to **always** return a python storage object to the user. Previously, we would error loudly if we detected that the storage had a null dataptr.

Instead, I updated the python bindings for the python storage methods that I saw involve data access, to throw an error later, only if you try to access those methods (e.g. `storage.data_ptr()` will now raise an error if the data ptr is null).




[ghstack-poisoned]

albanD commented on torch/csrc/DynamicTypes.cpp:
```diff
@@ -80,16 +80,14 @@ THPLayout* getTHPLayout(at::Layout layout) {
+  // When a user creates a python tensor wrapper subclass, the subclass
+  // is a tensor object that has a nullptr storage.
+  // We still allow users to call `my_subclass.untyped_storage()`, and get back
+  // a valid storage object (this can be useful for detecing aliasing
```
```suggestion
  // a valid storage object (this can be useful for detecting aliasing
```

albanD reviewed (approved):
SGTM!

bdhirsh committed:
Update on "Allow storage() to work on python tensor subclasses, but error on future data accesses"

This was discussed in feedback from the original version of my "reorder proxy/fake" PR. This PR allows calls to `tensor.untyped_storage()` to **always** return a python storage object to the user. Previously, we would error loudly if we detected that the storage had a null dataptr.

Instead, I updated the python bindings for the python storage methods that I saw involve data access, to throw an error later, only if you try to access those methods (e.g. `storage.data_ptr()` will now raise an error if the data ptr is null).




[ghstack-poisoned]

zou3519 commented on torch/csrc/StorageMethods.cpp:
```diff
@@ -103,6 +117,12 @@ static PyObject* THPStorage_new(PyObject* self, PyObject* noargs) {
+      storage.device_type() != c10::DeviceType::Meta &&
+      storage.sym_nbytes() != 0;
+  TORCH_CHECK(
+      !invalid, "Attempted to call resize() on an invalid python storage.")
```
nit: `resize_`. Also` fill_ `bellow.

bdhirsh committed:
Update on "Allow storage() to work on python tensor subclasses, but error on future data accesses"

This was discussed in feedback from the original version of my "reorder proxy/fake" PR. This PR allows calls to `tensor.untyped_storage()` to **always** return a python storage object to the user. Previously, we would error loudly if we detected that the storage had a null dataptr.

Instead, I updated the python bindings for the python storage methods that I saw involve data access, to throw an error later, only if you try to access those methods (e.g. `storage.data_ptr()` will now raise an error if the data ptr is null).




[ghstack-poisoned]
